### DISCUSSION TITLE
Determining Return Code and Testing for Network Errors in CURL

### DISCUSSION THREAD

**Discussion by xiaochongzi630** [curl/curl#12993] [Q&A]
When using CURL to communicate with a server, what return code is returned if the transmission completes successfully but a network error occurs during reception?
Also, how to test this scenario?

**Comment by jay** [comment]
> what return code is returned if the transmission completes successfully but a network error occurs during reception?

The curl command line tool and libcurl return [CURLE_OK](https://curl.se/libcurl/c/libcurl-errors.html) (0) on success. Can you please give an example of the scenario you are thinking of?

**Reply by xiaochongzi630** [reply]
Hi jay,

Thank you for your response

I am executing Azure processes using the CURL command. When executing Azure processes from CURL, I believe the following steps for sending and receiving are being taken:

[A]: CURL execution request → [B]: Azure process initiation → [C]: Response from Azure → [D]: Receiving the response via CURL

If there is a network error and the process at [A] terminates abnormally, I believe CURL would return CURLE_COULDNT_CONNECT (7). Is it possible for processes [A], [B], and [C] to terminate normally while only process [D] encounters an error? If only process [D] encounters an error, what is the return code?

**Reply by jay** [reply]
Can you disambiguate what you mean by processes; are A and D separate executions of the curl program or do you mean one execution with a request A and response D?

The scenario you are describing is curl returns an error while receiving the response. The error depends on how the transfer failed and is not guaranteed to always be the same code. For example a common network error code receiving data is CURLE_RECV_ERROR (56) but it could be any error code including CURLE_SEND_ERROR (55), CURLE_HTTP2 (16) etc. If curl is sending protocol data at the time it is receiving the response then that's how CURLE_SEND_ERROR could happen. See [libcurl-errors](https://curl.se/libcurl/c/libcurl-errors.html) error codes that start with CURLE_.

If you need more information about the transfer try [--write-out](https://curl.se/docs/manpage.html#-w) variables.

**Reply by xiaochongzi630** [reply]
Hi jay,

I am referring to one execution involving request A and response D. 

We are executing Azure processes using CURL, and in case of an error, we want to incorporate retry logic. 
However, we need to verify because if the Azure process has already been executed, retrying it could lead to undesired outcomes. 

Specifically, if a CURLE_COULDNT_CONNECT (7) error occurs, we want to confirm whether communication with Azure has truly failed. 
If communication has indeed failed, implying that the Azure process has not been executed yet, retrying should be safe. 
However, if there's a possibility that only process D has failed as mentioned above, there's a concern that retrying might lead to the Azure process being executed again, hence the need for caution.

CURLE_COULDNT_CONNECT (7) error occurs in what kind of situations?

**Reply by jay** [reply]
You would have to query the server in a follow up request. If there's no good way to do that then you could make an educated guess using write-out variables. For example check to see whether you received the expected response code via `%{response_code}`. However just because a response code has not been received (eg transfer failed after request was sent but before response) does not mean the server has not processed the request.

> CURLE_COULDNT_CONNECT (7) error occurs in what kind of situations?

Refer to the descriptions I linked to earlier: _Failed to connect() to host or proxy._

**Reply by xiaochongzi630** [reply]
Thanks for your response.

I understood that if a CURLE_COULDNT_CONNECT (7) error occurs, it means we can't determine whether the server has processed the request or not.

**Reply by jay** [reply]
If the connection could not be established then the request has not been sent.

(edit: some exceptions where the request is sent but must be sent again according to the server, in this case it technically has been sent. for example if [--location](https://curl.se/docs/manpage.html#-L) is used a connection may be established, the request sent, but the server responds with a redirect to a different location to send the request that curl cannot connect to, it will return CURLE_COULDNT_CONNECT in that case. if [--retry](https://curl.se/docs/manpage.html#--retry) is used and the server failed the request according to whatever retry reasons you set then the connection may be reestablished and if that fails curl will return CURLE_COULDNT_CONNECT.)

**Reply by xiaochongzi630** [reply]
Is it possible to determine whether the connection has been established?

**Reply by jay** [reply]
You can use [--write-out](https://curl.se/docs/manpage.html#-w) time variables. Internally those variables map to libcurl's [TIMES](https://curl.se/libcurl/c/curl_easy_getinfo.html#TIMES). Consider viewing that documentation as well for more detail. (eg [time_connect](https://curl.se/docs/manpage.html#timeconnect) maps to [CURLINFO_CONNECT_TIME_T](https://curl.se/libcurl/c/CURLINFO_CONNECT_TIME_T.html), [time_starttransfer](https://curl.se/docs/manpage.html#timestarttransfer) maps to [CURLINFO_STARTTRANSFER_TIME_T](https://curl.se/libcurl/c/CURLINFO_STARTTRANSFER_TIME_T.html)).

**Comment by dfandrich** [comment]
"Established" how? At the socket level? TCP level? TLS level? HTTP level?  Application level? It's not a straightforward answer.
